tracing/trace_keyspace_helper.cc

```diff
@@ -289,10 +289,23 @@ cql3::query_options trace_keyspace_helper::make_slow_query_mutation_data(const o
     const session_record& record = session_records.session_rec;
     auto millis_since_epoch = std::chrono::duration_cast<std::chrono::milliseconds>(record.started_at.time_since_epoch()).count();
 
-    // query command is stored on a parameters map with a 'query' key
+    sstring query_str;
+    // query command is stored on a parameters map with a 'query' key or 'query[N]' keys
     auto query_str_it = record.parameters.find("query");
     if (query_str_it == record.parameters.end()) {
-        throw std::logic_error("No \"query\" parameter set for a session requesting a slow_query_log record");
+        for (int i = 0; /* nothing */; ++i) {
+            // Extract the values under "query[0]", "query[1]", etc. in case of batched prepared statements
+            auto ith_query_iter = record.parameters.find(format("query[{}]", i));
+            if (ith_query_iter == record.parameters.end()) {
+                break;
+            }
+            query_str += ith_query_iter->second + " ";
```

Comment:
What is the maximum expected size of a batch statement? Can it be, say, a few thousand? I'm asking because a few thousand string allocations can be a strain on the memory allocator, and each string's `operator+=` would allocate. In that case, it would perhaps make sense to use a string builder (e.g. stringstream) instead of adding the strings one by one.

Also, I assume a common case is that we have a single `"query"` param instead of a batch, in which case we should perhaps elide allocations altogether instead of always creating a new string with a trailing space - so perhaps a single query deserves a separate code path which just forwards the string without new allocations.

Slow query tracing is not in any way on the fast path, but on the other hand it's often ran on strained clusters in order to investigate what's wrong with them, so I think it makes sense to minimize the overhead here.

Reply:
We've seen batches of tens of thousands so definitely worth doing.

Reply:
If it can be tens of thousands, then one more thing to consider on top of my previous comments is whether we want to construct the whole string... my point being that 50k requests, 100B each will create a 5MB string. The question is whether we need the whole string to investigate the error or maybe a shorter summary would suffice - e.g. a couple of first params and the total number of statements. I don't know the answer to that question, maybe someone from the field has an opinion. Anyway, it's probably a topic for a follow-up later, maybe worth leaving a TODO in the code.

Reply:
(by the way, processing tens of thousands strings in a non-preemptive loop is most likely a reactor stall candidate)